source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,266 @@
+}
+
+bool BufferedIoSocketHandleImpl::isOpen() const { return !closed_; }
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::readv(uint64_t max_length,
+                                                          Buffer::RawSlice* slices,
+                                                          uint64_t num_slice) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (owned_buffer_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  } else {
+    absl::FixedArray<iovec> iov(num_slice);
+    uint64_t num_slices_to_read = 0;
+    uint64_t bytes_to_read = 0;
+    for (; num_slices_to_read < num_slice && bytes_to_read < max_length; num_slices_to_read++) {
+      auto max_bytes_to_read =
+          std::min(std::min(owned_buffer_.length(), max_length) - bytes_to_read,
+                   uint64_t(slices[num_slices_to_read].len_));
+      owned_buffer_.copyOut(bytes_to_read, max_bytes_to_read, slices[num_slices_to_read].mem_);
+      bytes_to_read += max_bytes_to_read;
+    }
+    ASSERT(bytes_to_read <= max_length);
+    owned_buffer_.drain(bytes_to_read);
+    ENVOY_LOG(trace, "socket {} readv {} bytes", static_cast<void*>(this), bytes_to_read);
+    return {bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+  }
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                         uint64_t max_length) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (owned_buffer_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  } else {
+    // TODO(lambdai): Move at slice boundary to move to reduce the copy.
+    uint64_t max_bytes_to_read = std::min(max_length, owned_buffer_.length());
+    buffer.move(owned_buffer_, max_bytes_to_read);
+    return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+  }
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::writev(const Buffer::RawSlice* slices,
+                                                           uint64_t num_slice) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (!writable_peer_) {
+    return sysCallResultToIoCallResult(Api::SysCallSizeResult{-1, SOCKET_ERROR_INVAL});
+  }
+  if (writable_peer_->isWriteEndSet() || !writable_peer_->isWritable()) {
```

Comment:
I think these are wholly different conditions, aren't they? `writable_peer_->isWriteEndSet() == true` means that we have called shutdown(..._WR) on the local socket. And in this case the return value should EPIPE.

Reply:
Thanks for the explanation! That's a subtle error details. Fixing...

Reply:
fixed